polkadot/runtime/parachains/src/paras/mod.rs

```diff
@@ -506,6 +512,51 @@ impl OnNewHead for Tuple {
 	}
 }
 
+pub trait OnCodeUpgrade {
+	/// A function to execute some custom logic once the pre-checking is successfully completed.
+	///
+	/// This is currently used by the registrar pallet to perform refunds upon validation code
+	/// size reduction.
+	fn on_code_upgrade(id: ParaId) -> Weight;
+}
+
+/// An empty implementation of the trait where there is no logic executed upon a successful
+/// code upgrade.
+impl OnCodeUpgrade for () {
+	fn on_code_upgrade(_id: ParaId) -> Weight {
+		Weight::zero()
+	}
+}
+
+pub trait PreCodeUpgrade {
+	/// A function that performs custom logic to before sceduling a code upgrade.
+	///
+	/// This is currently utilized by the registrar pallet to ensure that the necessary validation
+	/// code upgrade costs are covered.
+	///
+	/// `requirements` signals whether to enforce the pre code upgrade requirements.
+	///
+	/// As a result, it indicates either the success or failure of executing the pre code upgrade
+	/// scheduling logic. In both cases, it returns the consumed weight.
+	fn pre_code_upgrade(
+		id: ParaId,
+		new_code: ValidationCode,
+		requirements: UpgradeRequirements,
```

Comment:
```suggestion
	fn can_pay_for_code_upgrade(
		id: ParaId,
		new_code: &ValidationCode,
```

Making it a little bit more explicit and then we can remove the `requirements` stuff. If we don't care about the payment (when done by root), we can just skip the check.

Reply:
But this isn't just performing checks, it is also updating the state regarding deposits of the parachain in the registrar. If we were to skip this when called by root the state wouldn't get properly updated.

Reply:
Hmm. I mean makes sense, but also in general you would just get back with the next runtime upgrade. (the deposit)